test/common/upstream/hds_test.cc

```diff
@@ -0,0 +1,288 @@
+
+namespace Envoy {
+namespace Upstream {
+
+// Friend class of HdsDelegate, making it easier to access private fields
+class HdsDelegateFriend {
+public:
+  // Allows access to private function processMessage
+  void processPrivateMessage(
+      std::shared_ptr<HdsDelegate> hd,
+      std::unique_ptr<envoy::service::discovery::v2::HealthCheckSpecifier>&& message) {
+    hd->processMessage(std::move(message));
+  };
+};
+
+class HdsTest : public testing::Test {
+public:
+  HdsTest()
+      : retry_timer_(new Event::MockTimer()), server_response_timer_(new Event::MockTimer()),
+        async_client_(new Grpc::MockAsyncClient()) {
+    node_.set_id("foo");
+  }
+
+  // Creates an HdsDelegate
+  void createHdsDelegate() {
+    InSequence s;
+    EXPECT_CALL(dispatcher_, createTimer_(_)).WillOnce(Invoke([this](Event::TimerCb timer_cb) {
+      retry_timer_cb_ = timer_cb;
+      return retry_timer_;
+    }));
+    EXPECT_CALL(dispatcher_, createTimer_(_))
+        .Times(AtLeast(1))
+        .WillOnce(Invoke([this](Event::TimerCb timer_cb) {
+          server_response_timer_cb_ = timer_cb;
+          return server_response_timer_;
+        }));
+    hds_delegate_.reset(new HdsDelegate(node_, stats_store_, Grpc::AsyncClientPtr(async_client_),
+                                        dispatcher_, runtime_, stats_store_, ssl_context_manager_,
+                                        secret_manager_, random_, test_factory_, log_manager_));
+  }
+
+  // Creates a HealthCheckSpecifier message that contains one endpoint and one
+  // healthcheck
+  envoy::service::discovery::v2::HealthCheckSpecifier* createSimpleMessage() {
+
+    envoy::service::discovery::v2::HealthCheckSpecifier* msg =
+        new envoy::service::discovery::v2::HealthCheckSpecifier;
+    msg->mutable_interval()->set_seconds(1);
+
+    auto* health_check = msg->add_health_check();
+    health_check->set_cluster_name("anna");
+    health_check->add_health_checks()->mutable_timeout()->set_seconds(1);
+    health_check->mutable_health_checks(0)->mutable_interval()->set_seconds(1);
+    health_check->mutable_health_checks(0)->mutable_unhealthy_threshold()->set_value(2);
+    health_check->mutable_health_checks(0)->mutable_healthy_threshold()->set_value(2);
+    health_check->mutable_health_checks(0)->mutable_grpc_health_check();
+    health_check->mutable_health_checks(0)->mutable_http_health_check()->set_use_http2(false);
+    health_check->mutable_health_checks(0)->mutable_http_health_check()->set_path("/healthcheck");
+
+    auto* socket_address =
+        health_check->add_endpoints()->add_endpoints()->mutable_address()->mutable_socket_address();
+    socket_address->set_address("127.0.0.0");
+    socket_address->set_port_value(1234);
+
+    return msg;
+  }
+
+  envoy::api::v2::core::Node node_;
+  Event::MockDispatcher dispatcher_;
+  Stats::IsolatedStoreImpl stats_store_;
+  MockClusterInfoFactory test_factory_;
+
+  std::shared_ptr<Upstream::HdsDelegate> hds_delegate_;
```

Comment:
Does this really need to be a shared_ptr? Could it just be unique? Shared pointers should only be used when non-trivial ownership semantics are required.